File: lib/product/incremental_cache.ml
type t = {
  schema : string;
  key : string;
  exit_code : int;
  report : string;
  integrity : string;
}

let key ~tool_version ~config_digest ~lock_digest inputs =
  let inputs =
    inputs
    |> List.map (fun (path, digest) -> (Util.normalize_slashes path, digest))
    |> List.sort (fun (left, _) (right, _) -> String.compare left right)
  in
  let material =
    Json.Object
      [
        ("config_digest", Json.String config_digest);
        ( "inputs",
          Json.Array
            (List.map
               (fun (path, digest) ->
                 Json.Object
                   [
                     ("digest", Json.String digest); ("path", Json.String path);
                   ])
               inputs) );
        ("lock_digest", Json.String lock_digest);
        ("schema", Json.String "analysis-cache-key-v1");
        ("tool_version", Json.String tool_version);
      ]
  in
  "sha256:" ^ Sha256.digest_string (Json.to_string material)

let unsigned_fields entry =
  [
    ("exit_code", Json.Int entry.exit_code);
    ("key", Json.String entry.key);
    ("report", Json.String entry.report);
    ("schema", Json.String entry.schema);
  ]

let unsigned_json entry = Json.Object (unsigned_fields entry)

let create ~key ~exit_code ~report =
  if exit_code < 0 || exit_code > 5 then
    Error "cache exit code must be 0..5"
  else
    let provisional =
      { schema = "analysis-cache-v1"; key; exit_code; report; integrity = "" }
    in
    let integrity =
      "sha256:"
      ^ Sha256.digest_string (Json.to_string (unsigned_json provisional))
    in
    Ok { provisional with integrity }

let to_json entry =
  Json.Object
    (("integrity", Json.String entry.integrity) :: unsigned_fields entry)

let to_canonical_json entry = Json.to_string (to_json entry) ^ "\n"

let required name converter json =
  match Option.bind (Json.member name json) converter with
  | Some value -> Ok value
  | None -> Error ("analysis cache needs field " ^ name)

let parse source =
  let open Util in
  let* json =
    match Json.parse source with
    | Ok value -> Ok value
    | Error error ->
        Error (Printf.sprintf "JSON byte %d: %s" error.offset error.message)
  in
  let* schema = required "schema" Json.as_string json in
  if schema <> "analysis-cache-v1" then
    Error ("unsupported analysis cache schema " ^ schema)
  else
    let* key = required "key" Json.as_string json in
    let* exit_code = required "exit_code" Json.as_int json in
    let* report = required "report" Json.as_string json in
    let* integrity = required "integrity" Json.as_string json in
    if exit_code < 0 || exit_code > 5 then Error "cache exit code must be 0..5"
    else
      let* rebuilt = create ~key ~exit_code ~report in
      if rebuilt.integrity <> integrity then
        Error "analysis cache integrity mismatch"
      else Ok rebuilt

File: crates/product/src/incremental_cache.rs
use std::collections::BTreeMap;
use workflow_verifier_foundation::{JsonValue, content_digest, normalize_slashes};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CacheKeyInput {
    pub path: String,
    pub digest: String,
}

#[must_use]
pub fn cache_key(
    tool_version: &str,
    config_digest: &str,
    lock_digest: &str,
    inputs: &[CacheKeyInput],
) -> String {
    let mut inputs: Vec<_> = inputs
        .iter()
        .map(|input| CacheKeyInput {
            path: normalize_slashes(&input.path),
            digest: input.digest.clone(),
        })
        .collect();
    inputs.sort_by(|left, right| left.path.cmp(&right.path));
    let material = JsonValue::Object(BTreeMap::from([
        (
            "config_digest".to_owned(),
            JsonValue::String(config_digest.to_owned()),
        ),
        (
            "inputs".to_owned(),
            JsonValue::Array(
                inputs
                    .iter()
                    .map(|input| {
                        JsonValue::Object(BTreeMap::from([
                            ("digest".to_owned(), JsonValue::String(input.digest.clone())),
                            ("path".to_owned(), JsonValue::String(input.path.clone())),
                        ]))
                    })
                    .collect(),
            ),
        ),
        (
            "lock_digest".to_owned(),
            JsonValue::String(lock_digest.to_owned()),
        ),
        (
            "schema".to_owned(),
            JsonValue::String("analysis-cache-key-v1".to_owned()),
        ),
        (
            "tool_version".to_owned(),
            JsonValue::String(tool_version.to_owned()),
        ),
    ]));
    content_digest(material.canonical())
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AnalysisCacheEntry {
    pub key: String,
    pub exit_code: i64,
    pub report: String,
    pub integrity: String,
}

impl AnalysisCacheEntry {
    /// Construct an authenticated analysis-cache-v1 entry.
    ///
    /// # Errors
    /// Rejects exit codes outside the public 0..=5 protocol range.
    pub fn new(
        key: impl Into<String>,
        exit_code: i64,
        report: impl Into<String>,
    ) -> Result<Self, String> {
        if !(0..=5).contains(&exit_code) {
            return Err("cache exit code must be 0..5".to_owned());
        }
        let mut entry = Self {
            key: key.into(),
            exit_code,
            report: report.into(),
            integrity: String::new(),
        };
        entry.integrity = content_digest(entry.unsigned_json().canonical());
        Ok(entry)
    }

    fn unsigned_fields(&self) -> BTreeMap<String, JsonValue> {
        BTreeMap::from([
            ("exit_code".to_owned(), JsonValue::Integer(self.exit_code)),
            ("key".to_owned(), JsonValue::String(self.key.clone())),
            ("report".to_owned(), JsonValue::String(self.report.clone())),
            (
                "schema".to_owned(),
                JsonValue::String("analysis-cache-v1".to_owned()),
            ),
        ])
    }

    fn unsigned_json(&self) -> JsonValue {
        JsonValue::Object(self.unsigned_fields())
    }

    #[must_use]
    pub fn verify_integrity(&self) -> bool {
        self.integrity == content_digest(self.unsigned_json().canonical())
    }

    #[must_use]
    pub fn to_canonical_json(&self) -> String {
        let mut fields = self.unsigned_fields();
        fields.insert(
            "integrity".to_owned(),
            JsonValue::String(self.integrity.clone()),
        );
        JsonValue::Object(fields).canonical_line()
    }

    /// Parse and authenticate strict analysis-cache-v1 product JSON.
    ///
    /// # Errors
    /// Rejects malformed/unknown fields, invalid exits, and integrity tampering.
    pub fn parse(source: &str) -> Result<Self, String> {
        let root = JsonValue::parse(source).map_err(|error| error.to_string())?;
        let fields = root.exact_object(
            "analysis cache",
            &["exit_code", "integrity", "key", "report", "schema"],
        )?;
        if fields.len() != 5 {
            return Err("analysis cache is missing a required field".to_owned());
        }
        let string = |name: &str| {
            fields
                .get(name)
                .and_then(JsonValue::as_str)
                .map(str::to_owned)
                .ok_or_else(|| format!("analysis cache needs string field {name}"))
        };
        if string("schema")? != "analysis-cache-v1" {
            return Err("unsupported analysis cache schema".to_owned());
        }
        let exit_code = fields
            .get("exit_code")
            .and_then(JsonValue::as_i64)
            .ok_or_else(|| "analysis cache needs integer field exit_code".to_owned())?;
        let supplied = string("integrity")?;
        let rebuilt = Self::new(string("key")?, exit_code, string("report")?)?;
        if rebuilt.integrity != supplied {
            return Err("analysis cache integrity mismatch".to_owned());
        }
        Ok(rebuilt)
    }
}
